### PR TITLE
ARROW-9225: [C++][Compute] Speed up counting sort

### PR DESCRIPTION
**Pull request by cyb70289**
For counting sort, using BitmapReader for null checking performs much
better than BitBlockCount.
This patch also drops counting nulls as it's not necessary.

### CODE REVIEW

**Comment by cyb70289** [comment]
**gcc-7.5, Intel Gold 5218**

```
                                               benchmark         baseline        contender  change %
9        SortToIndicesInt64Count/32768/10/min_time:1.000  391.232 MiB/sec  945.147 MiB/sec   141.582 {'null_percent': 10.0}
3         SortToIndicesInt64Count/32768/2/min_time:1.000  336.469 MiB/sec  617.053 MiB/sec    83.391 {'null_percent': 50.0}
13      SortToIndicesInt64Count/32768/100/min_time:1.000  539.575 MiB/sec  945.134 MiB/sec    75.163,{'null_percent': 1.0}
14        SortToIndicesInt64Count/32768/0/min_time:1.000  912.462 MiB/sec    1.061 GiB/sec    19.020,{'null_percent': 0.0}
12    SortToIndicesInt64Count/32768/10000/min_time:1.000  874.597 MiB/sec  948.007 MiB/sec     8.394 {'null_percent': 0.01}
0      SortToIndicesInt64Compare/32768/10/min_time:1.000  124.299 MiB/sec  133.614 MiB/sec     7.494 {'null_percent': 10.0}
4         SortToIndicesInt64Count/32768/1/min_time:1.000    4.212 GiB/sec    4.441 GiB/sec     5.455 {'null_percent': 100.0}
1     SortToIndicesInt64Compare/1048576/1/min_time:1.000    4.168 GiB/sec    4.380 GiB/sec     5.103 {'null_percent': 100.0}
6       SortToIndicesInt64Compare/32768/1/min_time:1.000    4.263 GiB/sec    4.428 GiB/sec     3.885 {'null_percent': 100.0}
8     SortToIndicesInt64Compare/32768/100/min_time:1.000  121.535 MiB/sec  126.154 MiB/sec     3.800 {'null_percent': 1.0}
11    SortToIndicesInt64Compare/8388608/1/min_time:1.000    3.833 GiB/sec    3.939 GiB/sec     2.746 {'null_percent': 100.0}
15      SortToIndicesInt64Count/1048576/1/min_time:1.000    4.189 GiB/sec    4.280 GiB/sec     2.175 {'null_percent': 100.0}
10      SortToIndicesInt64Compare/32768/0/min_time:1.000  127.065 MiB/sec  129.514 MiB/sec     1.927 {'null_percent': 0.0}
2       SortToIndicesInt64Compare/32768/2/min_time:1.000  211.863 MiB/sec  212.384 MiB/sec     0.246 {'null_percent': 50.0}
7       SortToIndicesInt64Count/8388608/1/min_time:1.000    3.881 GiB/sec    3.877 GiB/sec    -0.117 {'null_percent': 100.0}
5   SortToIndicesInt64Compare/32768/10000/min_time:1.000  124.847 MiB/sec  123.486 MiB/sec    -1.090 {'null_percent': 0.01}
```

**clang-9, Intel Gold 5218**
```
                                               benchmark         baseline         contender  change %
4        SortToIndicesInt64Count/32768/10/min_time:1.000  419.179 MiB/sec  1019.007 MiB/sec   143.096 {'null_percent': 10.0}
8         SortToIndicesInt64Count/32768/2/min_time:1.000  355.481 MiB/sec   658.370 MiB/sec    85.206 {'null_percent': 50.0}
5       SortToIndicesInt64Count/32768/100/min_time:1.000  574.780 MiB/sec     1.018 GiB/sec    81.451 {'null_percent': 1.0}
2         SortToIndicesInt64Count/32768/0/min_time:1.000  980.677 MiB/sec     1.137 GiB/sec    18.725 {'null_percent': 0.0}
15    SortToIndicesInt64Count/32768/10000/min_time:1.000  939.165 MiB/sec     1.023 GiB/sec    11.535 {'null_percent': 0.01}
13     SortToIndicesInt64Compare/32768/10/min_time:1.000  137.920 MiB/sec   152.108 MiB/sec    10.288 {'null_percent': 10.0}
10      SortToIndicesInt64Compare/32768/2/min_time:1.000  221.024 MiB/sec   243.047 MiB/sec     9.964 {'null_percent': 50.0}
3     SortToIndicesInt64Compare/32768/100/min_time:1.000  132.173 MiB/sec   141.416 MiB/sec     6.993 {'null_percent': 1.0}
12      SortToIndicesInt64Compare/32768/0/min_time:1.000  141.528 MiB/sec   148.206 MiB/sec     4.719 {'null_percent': 0.0}
7   SortToIndicesInt64Compare/32768/10000/min_time:1.000  135.794 MiB/sec   140.914 MiB/sec     3.770 {'null_percent': 0.01}
9       SortToIndicesInt64Count/8388608/1/min_time:1.000    3.683 GiB/sec     3.698 GiB/sec     0.421 {'null_percent': 100.0}
1     SortToIndicesInt64Compare/8388608/1/min_time:1.000    3.694 GiB/sec     3.651 GiB/sec    -1.159 {'null_percent': 100.0}
0     SortToIndicesInt64Compare/1048576/1/min_time:1.000    3.999 GiB/sec     3.951 GiB/sec    -1.198 {'null_percent': 100.0}
6       SortToIndicesInt64Count/1048576/1/min_time:1.000    3.999 GiB/sec     3.949 GiB/sec    -1.236 {'null_percent': 100.0}
11        SortToIndicesInt64Count/32768/1/min_time:1.000    4.096 GiB/sec     4.042 GiB/sec    -1.324 {'null_percent': 100.0}
14      SortToIndicesInt64Compare/32768/1/min_time:1.000    4.129 GiB/sec     4.018 GiB/sec    -2.686 {'null_percent': 100.0}
```

**Comment by cyb70289** [comment]
@ursabot benchmark --suite-filter=arrow-compute-vector-sort-benchmark --cc=clang-8 --cxx=clang++-8

**Comment by cyb70289** [comment]
@ursabot benchmark --suite-filter=arrow-compute-vector-sort-benchmark

File: cpp/src/arrow/compute/kernels/vector_sort.cc [hunk 1]
```diff
@@ -88,6 +88,28 @@ struct PartitionIndices {
   }
 };
 
+template <typename ArrayType, typename VisitorNotNull, typename VisitorNull>
+inline void VisitRawValueInline(const ArrayType& values,
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_sort.cc:92]: "VisitRawValuesInline" (plural) would be better

**Comment by pitrou** [comment]
I see no significant performance difference here on an AMD Ryzen CPU.
The weird thing with these performance numbers is that the `null_percent=0` case should probably be faster, no?

**Comment by cyb70289** [comment]
> I see no significant performance difference here on an AMD Ryzen CPU.
> The weird thing with these performance numbers is that the `null_percent=0` case should probably be faster, no?

Per @wesm [comment ](https://github.com/apache/arrow/pull/7525#issuecomment-648279829), he also didn't see significant performance difference after replacing bitmapreader with bitblockcount.

But I see big differences on AMD EYPC 7251, Intel Gold 5218, and Arm server. Both gcc and clang. No idea what's happening.

File: cpp/src/arrow/compute/kernels/vector_sort.cc (hunk 1 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/vector_sort.cc]: done

**Comment by cyb70289** [comment]
About `null_percent=0` case, the benchmark result I attached shows about 19%  improvement. Looks reasonable.

**Comment by wesm** [comment]
I'm interested to see if we could add a `BitmapReader::Advance` method so that we could employ a hybrid approach to use both BitBlockCounter and BitmapReader to get the best of both worlds

**Comment by wesm** [comment]
Here's my benchmarks on i9-9960X

```
$ archery benchmark diff --cc=gcc-8 --cxx=g++-8 cyb70289/sort master --suite-filter=vector-sort
                                               benchmark         baseline        contender  change %                                       counters
12       SortToIndicesInt64Count/32768/10/min_time:1.000    1.233 GiB/sec    2.062 GiB/sec    67.239    {'iterations': 56417, 'null_percent': 10.0}
1       SortToIndicesInt64Count/32768/100/min_time:1.000    1.605 GiB/sec    2.033 GiB/sec    26.628     {'iterations': 73247, 'null_percent': 1.0}
3         SortToIndicesInt64Count/32768/0/min_time:1.000    2.438 GiB/sec    2.850 GiB/sec    16.911    {'iterations': 111533, 'null_percent': 0.0}
7         SortToIndicesInt64Count/32768/2/min_time:1.000  805.132 MiB/sec  854.275 MiB/sec     6.104    {'iterations': 36039, 'null_percent': 50.0}
11    SortToIndicesInt64Compare/32768/100/min_time:1.000  147.729 MiB/sec  149.201 MiB/sec     0.996      {'iterations': 6588, 'null_percent': 1.0}
2       SortToIndicesInt64Count/1048576/1/min_time:1.000    4.559 GiB/sec    4.580 GiB/sec     0.458    {'iterations': 6503, 'null_percent': 100.0}
5     SortToIndicesInt64Compare/1048576/1/min_time:1.000    4.554 GiB/sec    4.560 GiB/sec     0.136    {'iterations': 6577, 'null_percent': 100.0}
0   SortToIndicesInt64Compare/32768/10000/min_time:1.000  148.181 MiB/sec  148.368 MiB/sec     0.126     {'iterations': 6659, 'null_percent': 0.01}
9       SortToIndicesInt64Compare/32768/2/min_time:1.000  245.628 MiB/sec  245.887 MiB/sec     0.106    {'iterations': 11171, 'null_percent': 50.0}
6      SortToIndicesInt64Compare/32768/10/min_time:1.000  156.791 MiB/sec  156.841 MiB/sec     0.032     {'iterations': 7073, 'null_percent': 10.0}
4       SortToIndicesInt64Count/8388608/1/min_time:1.000    4.276 GiB/sec    4.269 GiB/sec    -0.155     {'iterations': 772, 'null_percent': 100.0}
10    SortToIndicesInt64Compare/8388608/1/min_time:1.000    4.286 GiB/sec    4.273 GiB/sec    -0.305     {'iterations': 769, 'null_percent': 100.0}
8       SortToIndicesInt64Compare/32768/1/min_time:1.000    4.677 GiB/sec    4.648 GiB/sec    -0.634  {'iterations': 209517, 'null_percent': 100.0}
14        SortToIndicesInt64Count/32768/1/min_time:1.000    4.661 GiB/sec    4.617 GiB/sec    -0.945  {'iterations': 214256, 'null_percent': 100.0}
15      SortToIndicesInt64Compare/32768/0/min_time:1.000  151.334 MiB/sec  147.025 MiB/sec    -2.847      {'iterations': 6660, 'null_percent': 0.0}
13    SortToIndicesInt64Count/32768/10000/min_time:1.000    2.312 GiB/sec    2.131 GiB/sec    -7.844    {'iterations': 91492, 'null_percent': 0.01}
```

and clang-11

```
                                               benchmark         baseline        contender  change %                                       counters
15        SortToIndicesInt64Count/32768/0/min_time:1.000    2.069 GiB/sec    3.490 GiB/sec    68.665     {'iterations': 95577, 'null_percent': 0.0}
13       SortToIndicesInt64Count/32768/10/min_time:1.000    1.288 GiB/sec    1.895 GiB/sec    47.099    {'iterations': 59504, 'null_percent': 10.0}
10      SortToIndicesInt64Count/32768/100/min_time:1.000    1.551 GiB/sec    1.993 GiB/sec    28.456     {'iterations': 71755, 'null_percent': 1.0}
2     SortToIndicesInt64Count/32768/10000/min_time:1.000    1.975 GiB/sec    2.069 GiB/sec     4.771    {'iterations': 87376, 'null_percent': 0.01}
3       SortToIndicesInt64Compare/32768/0/min_time:1.000  155.058 MiB/sec  158.211 MiB/sec     2.033      {'iterations': 6935, 'null_percent': 0.0}
9       SortToIndicesInt64Compare/32768/2/min_time:1.000  248.515 MiB/sec  252.036 MiB/sec     1.417    {'iterations': 11676, 'null_percent': 50.0}
0       SortToIndicesInt64Compare/32768/1/min_time:1.000    4.678 GiB/sec    4.735 GiB/sec     1.222  {'iterations': 214204, 'null_percent': 100.0}
8         SortToIndicesInt64Count/32768/1/min_time:1.000    4.677 GiB/sec    4.729 GiB/sec     1.110  {'iterations': 213435, 'null_percent': 100.0}
5   SortToIndicesInt64Compare/32768/10000/min_time:1.000  151.573 MiB/sec  151.744 MiB/sec     0.113     {'iterations': 6769, 'null_percent': 0.01}
1      SortToIndicesInt64Compare/32768/10/min_time:1.000  162.339 MiB/sec  162.226 MiB/sec    -0.069     {'iterations': 7284, 'null_percent': 10.0}
14    SortToIndicesInt64Compare/32768/100/min_time:1.000  151.879 MiB/sec  151.589 MiB/sec    -0.191      {'iterations': 6837, 'null_percent': 1.0}
7     SortToIndicesInt64Compare/1048576/1/min_time:1.000    4.570 GiB/sec    4.553 GiB/sec    -0.383    {'iterations': 6579, 'null_percent': 100.0}
6       SortToIndicesInt64Count/1048576/1/min_time:1.000    4.569 GiB/sec    4.550 GiB/sec    -0.407    {'iterations': 6459, 'null_percent': 100.0}
11      SortToIndicesInt64Count/8388608/1/min_time:1.000    4.280 GiB/sec    4.240 GiB/sec    -0.939     {'iterations': 771, 'null_percent': 100.0}
12    SortToIndicesInt64Compare/8388608/1/min_time:1.000    4.281 GiB/sec    4.222 GiB/sec    -1.369     {'iterations': 758, 'null_percent': 100.0}
4         SortToIndicesInt64Count/32768/2/min_time:1.000  835.964 MiB/sec  752.317 MiB/sec   -10.006    {'iterations': 37122, 'null_percent': 50.0}
```

+1, thanks @cyb70289!

**Pull request closed without merging**